### DISCUSSION TITLE
Automatic voltage using a physical switch

### DISCUSSION THREAD

**Discussion by Krystman** [sanni/cartreader#883] [Ideas & Suggestions]
Is there a reason not to use a physical switch to set the voltage automatically for GB vs GBA carts like in the real GBA? In the real GBA there is a small physical switch that is being pressed by the case of a GB cart that triggers 5v mode. If a GBA cart is inserted a groove in the casing will leave the switch open for 3v. I've seen the [GBA HD kit by Consoles4U](https://consoles4you.ch/en/gbahd-upcycle-shield-diy-kit) use [this specific component](https://de.aliexpress.com/item/1005005522737398.html?spm=a2g0o.productlist.main.45.774d7062sDliDv&algo_pvid=463f96e6-1b4e-4825-b953-ab103a51e161&algo_exp_id=463f96e6-1b4e-4825-b953-ab103a51e161-22&pdp_npi=4%40dis%21EUR%211.11%211.11%21%21%211.19%21%21%40210387dd17044783344496684e180c%2112000033406914650%21sea%21DE%21724000501%21&curPageLogUid=z0AFkKXNlr8K) to replicate the behavior.
<img width="411" alt="Screenshot 2024-01-05 175634" src="https://github.com/sanni/cartreader/assets/10092987/69a1e05b-2003-42cf-a46f-19b6e559db34">

<img width="226" alt="Screenshot 2024-01-05 192631" src="https://github.com/sanni/cartreader/assets/10092987/d65a5149-d482-4e14-863b-e4e079052dc0">

![image](https://github.com/sanni/cartreader/assets/10092987/384d7602-fac6-49a2-b805-902cecef7bc6)

**Comment by Ancyker** [comment]
There is automatic voltage selection via [VSELECT](https://github.com/sanni/cartreader/wiki/Automatic-Voltage-Selection) and the firmware. There isn't much to be gained from adding a physical switch to HW5 as you still need to select GBA in the menu.

**Reply by Krystman** [reply]
Yes, I know. And the VSELECT solution is obviously superior. But also it requires 5 parts and some of them are the size of sand grains. And this is just a simple switch.

**Reply by Ancyker** [reply]
Right, it's a switch that can tell the MCU if a GBA cart is inserted or not when you choose GB. What changes the voltage? Vselect would still be needed. All of this also assumes that we have free GPIO pins available, we don't.

With that said, I had this same idea long ago which is why it's part of [HW7](https://github.com/sanni/cartreader/discussions/874#discussioncomment-7733074) (and likely 6). So, it will come, just not to HW5.

**Reply by Ancyker** [reply]
I was thinking on this post a bit more and I think I misunderstood what you were suggesting. I think you wanted the switch itself to change the voltage going to the GB slot based on the switch position and leave the firmware completely out of it?

If so, it'd technically be a viable way on a GBx-only OSCR. However, there are a few issues with it. The biggest is that a switch alone is not enough, more components would be needed, at a minimum:

![image](https://github.com/sanni/cartreader/assets/7105333/527e3dba-35cc-4655-9b55-e6797319180e)

Q1 and Q2 would be connected to the switch, one through an inverting logic gate. If this image looks familiar, it's because it's from the datasheet for the TPS2113PW, the IC that Vselect uses. Yeah, we'd just be re-implementing it using discrete components. Some of these will need to be SMD components as well, so, now we are defeating the purpose of why you wanted the switch to begin with.

There is a lazier way to do this with fewer components but it's not elegant. You could run 3.3V through a diode directly to the VCC pin of the slot plus use a normally open switch connected to 5V on the other side of the diode. Now when a GBA cart is inserted 3.3V is supplied since the switch isn't depressed and 5V is cut off. When a GB/GBC cart is inserted the switch closes and allows 5V to flow to VCC while the diode prevents 5V from back feeding onto our 3V3 rail. This would work, but it's not great.

An even better way to do it would be to use isolated supplies and transistors to control the ground side. This is similar to cloning Vselect's IC but these types of transistors tend to be cheaper.

But once again this all has one big caveat attached to it: It only makes sense if the GBx slot is the only slot as you'd need to power the Arduino off of this same supply or it won't use the correct logic voltage. The only way to viably support this method and other systems would be to instruct the user to leave the voltage switch set to 5V when using the GB slot and then override 5V when a GBA cart is detected. This would require so many components you might as well just buy an assembled Vselect add-on board because it would be cheaper and less prone to failure. I'm also glossing over a lot of other design considerations.

And finally, while it's not recommended, the 0402 components aren't needed to make Vselect work. You could just install the TPS2113PW (though if doing that, I recommend going with the TPS2112PW). It may behave erratically if you do this though, I haven't tested that. But. as I understand it the main thing affected would be the amount of time it takes to switch voltages (it would increase). If that is indeed the only issue then this would only be a concern on initial power up as the default is 5V until the firmware boots and switches it to 3.3V. If the time to switch is increased this increases the risk of damaging 3.3V carts connected to the system during boot. But I already discourage leaving carts, especially 3.3V carts. connected during boot.

So yeah, in the end, this either just doesn't work in the case of GBx + other systems or does work but is so complex it's not viable in the case of a GBx-only system. Don't feel bad though, the reason I knew all of this offhand and may have seemed to dismiss the idea in a quick/casual manner is I already had this very same idea. In fact, it's part of why Vselect exists. I started here, using physical switches, before moving on to Vselect and firmware control. I point this out because I don't want you or anyone to feel discouraged when it comes to coming up with new suggestions/ideas.

I will say that with HW7 the only physical switch for cart detection that's planned right now is the one for the GBx. For the rest I chose a different route as adjusting the height of switches for vertical slots seemed too convoluted. Instead I'm currently looking at IR LEDs and IR photoresistors. The idea being that when a cart is inserted the 2 will be partially covered and thus the reflected IR will be more intense. The downside of this method over switches is it requires polling the sensors where as with switches interrupts could have been used. This is actually my next goal to get a proof of concept going for. I need to figure out how close together the 2 should be to work reliably, minimizing false triggers and maximizing detection rates.

**Reply by Krystman** [reply]
First of all, thank you for taking the time to write such a detailed and thoughtful answer. This is delightful and I am learning a lot.

I was indeed thinking of doing the switching physically and without involving the firmware. I'm suprised to hear that it would take so many more components considering the regular voltage switch is also just a physical switch. Is that because these two switches don't work the same? Or is it because we would have two switches controlling the voltage?

The point about this only really working for GBx is well taken. It is mainly the system I am interested in so I am a bit biased here.

Also, this is a bit OT but why are 0402 components used in the VSELECT build as opposed to 1206 for example? For a newcomer the difficulty jump between the two is considerable.

**Reply by Ancyker** [reply]
> Also, this is a bit OT but why are 0402 components used in the VSELECT build as opposed to 1206 for example? For a newcomer the difficulty jump between the two is considerable.

Put simply, board space. The components were initially chosen for the add-on board which had to fit where the voltage switch normally goes. Other factors can include ESR/ESL which are naturally lower with smaller packages. Cost could be yet another factor as 1206 capacitors will cost slightly more. However, in this case, these weren't really considered; it was primarily board space.

The integrated version just kept these packages for simplicity. This turned out to be fortuitous as it saved changing them back to 0402 later as 1206 would not fit on later revisions. It would have been possible (though a minor annoyance) to fit them on the current release revision, but not really reasonable on the next revision:

![image](https://github.com/sanni/cartreader/assets/7105333/3db526ea-f9e3-4134-a1d9-ba26ca2f06e8)
![image](https://github.com/sanni/cartreader/assets/7105333/bb1a7d1a-7e31-438a-b282-04fbcefed989)
